File: go/testing/harness.go
package testing

import (
	"context"
	"fmt"
)

// RunOptions configure RunDeliveries.
type RunOptions struct {
	Transport *MemoryTransport
	Scheduler *MemoryScheduler
	Clock     *TestClock
	// MaxRounds caps delivery rounds (defends against a non-converging
	// schedule). 0 means the default (200).
	MaxRounds int
}

// RunDeliveries repeatedly drains the transport, then jumps the clock to the
// next scheduled retry and releases it, until nothing is queued or
// scheduled. This is the in-memory analogue of a worker plus the durable
// scheduler firing over time. Returns an error if MaxRounds is exceeded.
func RunDeliveries(ctx context.Context, opts RunOptions) error {
	maxRounds := opts.MaxRounds
	if maxRounds <= 0 {
		maxRounds = 200
	}
	for round := 0; round < maxRounds; round++ {
		if _, err := opts.Transport.Drain(ctx); err != nil {
			return err
		}
		if opts.Scheduler.Size() == 0 {
			return nil
		}
		if next, ok := opts.Scheduler.NextDueAt(); ok {
			opts.Clock.Set(next)
		}
		if _, err := opts.Scheduler.Advance(ctx); err != nil {
			return err
		}
	}
	return fmt.Errorf("RunDeliveries exceeded maxRounds=%d (schedule not converging)", maxRounds)
}

File: packages/testing/src/harness.ts
/** @fileoverview Drives the in-memory delivery loop to quiescence for integration tests. @module @anyhook/testing */
import type { MemoryTransport } from './memory-transport.js';
import type { MemoryScheduler } from './memory-scheduler.js';
import type { TestClock } from './test-clock.js';

export interface RunOptions {
  transport: MemoryTransport;
  scheduler: MemoryScheduler;
  clock: TestClock;
  /** Safety cap on delivery rounds (defends against a non-converging schedule). */
  maxRounds?: number;
}

/**
 * Repeatedly drain the transport, then jump the clock to the next scheduled retry and
 * release it, until nothing is queued or scheduled. This is the in-memory analogue of a
 * worker + the durable scheduler firing over time. Throws if `maxRounds` is exceeded.
 */
export async function runDeliveries(opts: RunOptions): Promise<void> {
  const { transport, scheduler, clock, maxRounds = 200 } = opts;
  for (let round = 0; round < maxRounds; round++) {
    await transport.drain();
    if (scheduler.size === 0) return;
    const next = scheduler.nextDueAt();
    if (next !== undefined) clock.set(next);
    await scheduler.advance();
  }
  throw new Error(`runDeliveries exceeded maxRounds=${maxRounds} (schedule not converging)`);
}
